### ISSUE TITLE
hookspec for just doing something without providing a dock widget

### ISSUE THREAD

**Issue by haesleinhuepf** [napari/napari#3015]
## 🚀 Feature
Hi all,

I would like to request for another hookspec that takes functions which are executed when a menu entry is clicked. Examples could be "Reset Brightness & Contrast" or "Copy current layer".

## Motivation

I'd like to implement functions such as for resetting brightness contrast
```
def reset_brightness_contrast(image: Image):
    image.contrast_limits = (image.data.min(), image.data.max())
```

At the moment, the user has to confirm the action in a dock widget which appears a bit cumbersome:
![image](https://user-images.githubusercontent.com/12660498/125155037-956ba500-e15d-11eb-9fd5-fc5e30110158.png)

If the operation could just be executed on the current layer[s], that would be cooll. You find more examples [here](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_convert_to_numpy.py)

## Pitch

We could extend napari with easy-to-use single-click functions.

## Additional context

At some point, we may also start collecting single-click functions such as "Reset brightness & contrast" in a common repository.

**Comment by tlambert03** [comment]
yep, now that #2556 is in, we need to think about how to expose these take-a-layer-and-do-something functions externally.  This is all rather tightly connected to the concept of the action manager, which is coming along nicely as well.  More generally,  we can describe these single-click functions as callables that happen in response to an event (button click, mouse press, etc.) with some degree of contextual awareness (a specific layer type is selected, etc...)

We definitely plan on exposing such a mechanism, but probably need the action manager and context concept to mature just a bit.

Side note: "auto-scaling" definitely should just be easier to access in the gui.  It's the "reset" button in the contrast limits controller that you get when you right click (we agree it's not discoverable enough)

<img width="331" alt="Untitled" src="https://user-images.githubusercontent.com/1609449/125168483-c8b92e80-e173-11eb-9f3d-34e1adb8eb47.png">

programmatically, you can use `layer.reset_contrast_limits()`... so don't need a new function for that one (but if you want to add percentile to it, that would be a nice PR).

**Comment by sofroniewn** [comment]
> We definitely plan on exposing such a mechanism, but probably need the action manager and context concept to mature just a bit.

Yes, i think to me these will come under the auspices of the action manger. Look here @haesleinhuepf at the idea of a plugin registering actions that can be used as keybindings https://github.com/napari/napari-animation/pull/103

My hope is that we can have the same action registration idea for all actions, and then have a separate ability to wire them up, would you like them to appear in a menubar, a context menu, or a keyboard shortcut.

Would be great to get @Carreau thoughts on this too, also @nclack might want to track this convo too

**Comment by sofroniewn** [comment]
> My hope is that we can have the same action registration idea for all actions, and then have a separate ability to wire them up, would you like them to appear in a menubar, a context menu, or a keyboard shortcut.

I guess one open question is if we'd also have event connection run through the action manager / a provide action hookspec, or if they'd be their own hookspec.

i.e. if you could provide something like

```python
('napari.layers.Layer.events.opacity', my_opacity_callback)
# or possibly just
('napari.layers.Layer.opacity', my_opacity_callback)
```

And then when your plugin was active/ enabled we'd connect your callbacks (possibly gated on some other information too).

Curious what folks think of this idea too?

**Comment by haesleinhuepf** [comment]
I'm not sure if I communicated my initial suggestion well. I just would like to use a hook for generic functions that get executed on menu-click.  :-)

**Comment by tlambert03** [comment]
I think we did understand the gist of the request... but I'm not sure it's as simple as "just do this when I click".  there's a lot of contextual stuff to take into consideration.  That said, if you have a spec in mind, with an idea for how it might be implemented, feel free to propose!

**Comment by haesleinhuepf** [comment]
> That said, if you have a spec in mind, with an idea for how it might be implemented, feel free to propose!

Yeah, sure! I thought of exactly the same as `napari_experimental_provide_function`, but instead of adding the function-widget to the GUI, the function is just called. :-)

**Comment by tlambert03** [comment]
does the function take no inputs?

**Comment by tlambert03** [comment]
In my opinion, `napari_experimental_provide_function` is still kind of a "hook without an implementation".  We basically don't use it anywhere except to make a magicgui widget for the user (which wasn't exactly the original motivation for it)... and it's exactly because it's unclear exactly how it should be used (similar to the question I'm posing here)

**Comment by haesleinhuepf** [comment]
Just like `napari_experimental_provide_function` the functions can have inputs such as `napari_viewer : Viewer` or `image : Image`, `labels : Labels` or `layer : Layer`. It think that pretty much all of that is already implemented. I'm just not sure which image or layer should be passed in case no layer is selected. Maybe none should be passed then.

**Comment by tlambert03** [comment]
> It think that pretty much all of that is already implemented

the "implementation" there (i.e. the association of those type annotations with a python object that gets passed to the function) is entirely in magicgui... where a stateful connection between a user-mediated layer selection can be made _because_ of a widget.  That is, that function doesn't go looking for the currently selected layer in `viewer.layers` or anything.  it "knows" what layer to use only because of the magicgui widget that is provided to the use.  So we really don't have a widget-free implementation of those concepts yet

(edit: I'm not trying to suggest that any of this is terribly difficult... but it certainly isn't ready-and-waiting to just be turned on and exposed.  hence the need for a little discussion)

**Comment by haesleinhuepf** [comment]
Ah I see. Ok, so it's more complicated than I thought. 

Just another example of what I'm thinking of to draw a better picture: I have [this function](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_convert_to_numpy.py#L26-L27) which copies a labels layer:
```
def make_labels_editable(labels : Labels) -> Labels:
    return Labels(np.asarray(labels.data), name="np " + labels.name)
```

If a labels layer is selected, this could just be executed. No need to confirm a dialog ;-)

**Comment by tlambert03** [comment]
> If a labels layer is selected, this could just be executed

yep, agreed... but that "if a labels layer is selected" sentence there is exactly the sort of context I was referring to [above](https://github.com/napari/napari/issues/3015#issuecomment-877841437) when I said "there's a lot of contextual stuff to take into consideration".  In other words: yes, it's very simple to say "hey, is a label's layer selected?" and satisfy your use case above.  but what if someone wants their function to be called only if two images are selected?  what if they want an image and a labels?  what if they only want a 3D image, not a 2D image?  These are the sorts of contextual things we're trying to codify.

I definitely suggest taking a look through the concepts introduced in #2556, since I think that's a promising way forward here.  Specifically, have a look at the "available layerlist contexts" defined [here](https://github.com/napari/napari/blob/4cfe609c71bd5d9347e3c93da0bec20dd662768d/napari/components/layerlist.py#L346-L364), and how those context keys are used in the "`enable_when`" keys of these actions [here](https://github.com/napari/napari/blob/4cfe609c71bd5d9347e3c93da0bec20dd662768d/napari/layers/_layer_actions.py#L114-L173).  Those layer "actions" are the equivalent of your  `make_labels_editable` command there and the `enable_when` context allows us to know when it's a valid function to offer up to the user, and when it's not.  for instance, now if you click on an RGB image, the "split RGB" action is available:

<img width="347" alt="Untitled" src="https://user-images.githubusercontent.com/1609449/125206735-c085ef80-e256-11eb-82bf-c36d2c06ced2.png">

whereas if you have three grayscale images of the same shape selected, you get "merge to stack"

<img width="337" alt="Untitled3" src="https://user-images.githubusercontent.com/1609449/125206761-de535480-e256-11eb-8c6b-d17a13f41a2e.png">

make sense?   So, we're _close_ to being able to open up that sort of a "do this simple action if this simple condition is met"... but I do think it's a touch more complicated than at first it may appear

**Comment by haesleinhuepf** [comment]
Oh wow, I wasn't aware that all this awesome stuff is coming!! Maybe my request becomes obsolete then. I'll practice patience 😊
